presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java

```diff
@@ -0,0 +1,350 @@
+                    }
+                    else if (name.equals(destinationTable)) {
+                        return ImmutableList.of(
+                                new ColumnMetadata(destinationColumnNameA, INTEGER),
+                                new ColumnMetadata(destinationColumnNameB, INTEGER));
+                    }
+                    throw new IllegalArgumentException();
+                })
+                .withApplyTableScanRedirect(applyTableScanRedirect);
+        applyProjection.ifPresent(builder::withApplyProjection);
+        applyFilter.ifPresent(builder::withApplyFilter);
+
+        queryRunner.createCatalog(MOCK_CATALOG, builder.build(), ImmutableMap.of());
+        return queryRunner;
+    }
+
+    private Optional<ProjectionApplicationResult<ConnectorTableHandle>> mockApplyProjection(
+            ConnectorSession session,
+            ConnectorTableHandle tableHandle,
+            List<ConnectorExpression> projections,
+            Map<String, ColumnHandle> assignments)
+    {
+        MockConnectorTableHandle handle = (MockConnectorTableHandle) tableHandle;
+
+        List<ColumnHandle> newColumns = assignments.values().stream()
+                .collect(toImmutableList());
+        if (handle.getColumns().isPresent() && newColumns.equals(handle.getColumns().get())) {
+            return Optional.empty();
+        }
+
+        return Optional.of(
+                new ProjectionApplicationResult<>(
+                        new MockConnectorTableHandle(handle.getTableName(), handle.getConstraint(), Optional.of(newColumns)),
+                        projections,
+                        assignments.entrySet().stream()
+                                .map(assignment -> new Assignment(
+                                        assignment.getKey(),
+                                        assignment.getValue(),
+                                        ((MockConnectorColumnHandle) assignment.getValue()).getType()))
+                                .collect(toImmutableList())));
+    }
+
+    private Optional<ConstraintApplicationResult<ConnectorTableHandle>> mockApplyFilter(
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            Constraint constraint)
+    {
+        MockConnectorTableHandle handle = (MockConnectorTableHandle) table;
+
+        TupleDomain<ColumnHandle> oldDomain = handle.getConstraint();
+        TupleDomain<ColumnHandle> newDomain = oldDomain.intersect(constraint.getSummary());
+        if (oldDomain.equals(newDomain)) {
+            return Optional.empty();
+        }
+
+        return Optional.of(
+                new ConstraintApplicationResult<>(
+                        new MockConnectorTableHandle(handle.getTableName(), newDomain, Optional.empty()),
+                        TupleDomain.all()));
+    }
+
+    private Optional<TableScanRedirectApplicationResult> mockApplyRedirectAfterProjectionPushdown(
```

Comment:
remove, unused. You can use `getMockApplyRedirectAfterPredicatePushdown` instead

Reply:
`mockApplyRedirectAfterProjectionPushdown` applies redirection only if it sees projected columns.
`getMockApplyRedirectAfterPredicatePushdown` applies redirection only if it sees pushed down predicates.
I would prefer to not mix the two implementations.

Reply:
OK. I didn't notice it.